shaders/commonProfileShaders.cpp

```diff
@@ -55,11 +56,17 @@ namespace GLTF
         ((std::istream*)png_get_io_ptr(pngPtr))->read((char*)data, length);
     }
 #endif
+	static std::unordered_map<std::string, bool> haveAlphas;		// Caching to remove up to 98% execution time
```

Comment:
@elfprince13 Yes I recognize that hashing is of lower complexity than binary tree lookups.

The observation is that with string keys, binary tree lookups often take less runtime than hashing + comparison from collisions, even though the algorithmic complexity is theoretically greater. The memory footprint is also usually higher. These are especially true for small sets.

http://kariddi.blogspot.com/2012/07/c11-unorderedmap-vs-map.html

http://blog.dubbelboer.com/2012/12/04/lru-cache.html

That being said, your source is more recent than mine so it is possible these containers have been made more efficient in more recent versions of the standard libraries. I decided to try a little microbenchmark in MSVC++ 14.0.

I added the same set of randomly generated 32 character keys to a map and unordered_map and then timed 100000 random lookups from that key set into both.

| numKeys | map | unordered_map |
| --- | --- | --- |
| 1 | 425 ms | 605 ms |
| 2 | 544 ms | 701 ms |
| 3 | 566 ms | 798 ms |
| 4 | 573 ms | 790 ms |
| 5 | 611 ms | 716 ms |
| 6 | 638 ms | 694 ms |
| 7 | 667 ms | 717 ms |
| 8 | 688 ms | 700 ms |
| 9 | 702 ms | 608 ms |
| 10 | 720 ms | 613 ms |
| 11 | 750 ms | 653 ms |

_Windows 10, MSVC++ 14.0, Intel(R) Core(TM) i7-4700MQ CPU @ 2.40GHz_

And of course as we expect, our constant hashing time takes over after a certain threshold. At least on my machine, that threshold was around 8. However, for small sets of typically only a few elements, the map does seem to me to be the better choice.

I've included my benchmarking code below:
```c++
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <map>
#include <unordered_map>

std::string randomString(size_t length){
	auto randomChar = []() -> char
	{
		const char charset[] =
			"0123456789"
			"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
			"abcdefghijklmnopqrstuvwxyz";
		const size_t maxIndex = (sizeof(charset) - 1);
		return charset[rand() % maxIndex];
	};
	std::string str(length, 0);
	std::generate_n(str.begin(), length, randomChar);
	return str;
}

int main(){
	std::vector<std::string> keys;
	std::map<std::string, bool> orderedMap;
	std::unordered_map<std::string, bool> unorderedMap;

	int numKeys = $N; // replace this with the number of keys
	int keyLength = 32;

	for (int i = 0; i < numKeys; i++) {
		std::string key = randomString(keyLength);
		keys.push_back(key);
		orderedMap[key] = true;
		unorderedMap[key] = true;
	}

	std::clock_t startOrdered, endOrdered, startUnordered, endUnordered;
	startOrdered = std::clock();
	for (int i = 0; i < 100000; i++) {
		bool lookup = orderedMap[keys[rand() % numKeys]];
	}
	endOrdered = std::clock();

	startUnordered = std::clock();
	for (int i = 0; i < 100000; i++) {
		bool lookup = unorderedMap[keys[rand() % numKeys]];
	}
	endUnordered = std::clock();

	std::cout << "Ordered Search Time: " << (endOrdered - startOrdered) / (double)(CLOCKS_PER_SEC / 1000) << " ms" << std::endl;
	std::cout << "Unordered Search Time: " << (endUnordered - startUnordered) / (double)(CLOCKS_PER_SEC / 1000) << " ms" << std::endl;

    return 0;
}
```